server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -1614,6 +1618,33 @@ public void testReservesBytesDuringPeerRecoveryPhaseOne() throws Exception {
         );
     }
 
+    public void testPeerRecoveryTargetRetriesIfSourceRepliesWithCBE() throws Exception {
+        var sourceNode = internalCluster().startNode();
+        String indexName = randomIdentifier();
+        final int allocationMaxRetry = randomIntBetween(1, 5);
+        createIndex(
+            indexName,
+            Settings.builder()
+                .put(IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), 1)
+                .put(IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 0)
+                .put(MaxRetryAllocationDecider.SETTING_ALLOCATION_MAX_RETRY.getKey(), allocationMaxRetry)
+                .build()
+        );
+        ensureGreen(indexName);
+        MockTransportService transportService = (MockTransportService) internalCluster().getInstance(TransportService.class, sourceNode);
+        AtomicInteger startRecoveryAttempts = new AtomicInteger();
+        transportService.addRequestHandlingBehavior(PeerRecoverySourceService.Actions.START_RECOVERY, (handler, request, channel, task) -> {
+            if (startRecoveryAttempts.incrementAndGet() <= allocationMaxRetry) {
+                throw new CircuitBreakingException("Simulated circuit_breaking_exception", randomFrom(CircuitBreaker.Durability.values()));
```

Comment:
Just checking: does this do the appropriate amount of wrapping in `RemoteTransportException` and so on?

Also if the target node responds to the source node with a CBE then we retry up to some timeout (`indices.recovery.internal_action_retry_timeout` which is 60s by default) but after that timeout we propagate the CBE back to the target, wrapped in two `RemoteTransportException` wrappers, and with this change I think we'd retry the recovery from the start in that case too. That seems a little pointless, if we're going to retry forever we may as well not go back to the start.

Reply:
> does this do the appropriate amount of wrapping in RemoteTransportException and so on?

The target receives the CBE wrapped in `RemoteTransportException`, unwraps and based on the cause calls retry again.

> Also if the target node responds to the source node with a CBE then we retry up to ... That seems a little pointless, if we're going to retry forever we may as well not go back to the start.

Are you suggesting that instead of `recoverySettings.activityTimeout()` we should use a much higher value?

Reply:
Also as far as the retrying goes, since the peer recovery source here would be the primary, I guess there is no harm in retrying, since there is no where else to recover from, right? Meaning either we keep retrying or fail the shard?

Reply:
The issue I see with retrying forever is that there's only a limited number of recovery slots on each node. If this recovery isn't going to complete because of a perennially-circuit-breaking node then that will prevent other recoveries from completing too.

Reply:
ok. I see. I am not sure how to handle that! I agree it is not desirable. So is it safer NOT to follow up with this approach as long as we have the limited ongoing recoveries? Alternatively, we could limit the CBE-based retry to the one minute and not retry again after that. But I am not sure how much work that is and whether just a 1 min retry brings much of a value here.

Reply:
Yes I think we can't retry forever as things stand today. The trouble is that I see no particularly easy way to limit the retries on the target node either, there's nothing like that today, and I'm not sure it's worth introducing such a mechanism.

I'd like to get to a point where we _can_ retry forever without involving the master or blocking other recoveries forever, but that's a bigger change. We have discussed this elsewhere and I'll open an issue about it.

Reply:
I opened https://github.com/elastic/elasticsearch/issues/98087

Reply:
I guess, for now I'll abandon this. FWIW, the 1 min timeout also doesn't seem to kick in. My understanding of `RecoveryMonitor` is that it only fails the shard after the timeout if there are no activities (e.g. recovery stuck), otherwise continuously trying every 5s doesn't result in the timeout.